OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java

```diff
@@ -572,139 +443,177 @@ private static OSEmailSubscriptionState getLastEmailSubscriptionState(Context co
+    * @param newAppId - String app id associated with the OneSignal dashboard app
+    */
+   public static void setAppId(@NonNull String newAppId) {
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppId(id) called with app_id: " + newAppId + "!");
 
-      try {
-         ApplicationInfo ai = context.getPackageManager().getApplicationInfo(context.getPackageName(), PackageManager.GET_META_DATA);
-         Bundle bundle = ai.metaData;
+      if (newAppId == null || newAppId.isEmpty()) {
+         OneSignal.onesignalLog(LOG_LEVEL.WARN, "newAppId is null or empty, ignoring!");
+         return;
+      }
+      else if (!newAppId.equals(appId)) {
+         // Pre-check on app id to make sure init of SDK is performed properly
+         //     Usually when the app id is changed during runtime so that SDK is reinitialized properly
+         initDone = false;
+      }
 
-         String sender_id = bundle.getString("onesignal_google_project_number");
-         if (sender_id != null && sender_id.length() > 4)
-            sender_id = sender_id.substring(4);
+      appId = newAppId;
 
-         OneSignal.init(context, sender_id, bundle.getString("onesignal_app_id"), mInitBuilder.mNotificationOpenedHandler, mInitBuilder.mNotificationReceivedHandler);
-      } catch (Throwable t) {
-         t.printStackTrace();
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppId(id) finished, checking if appContext has been set before proceeding...");
+      if (appContext == null) {
+         OneSignal.onesignalLog(LOG_LEVEL.WARN, "appId set, but please call setAppContext(appContext) with Application context to complete OneSignal init!");
+         return;
+      }
+
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppId(id) successful and appContext is set, continuing OneSignal init...");
+      init();
+   }
+
+   /**
+    * 1/2 steps in OneSignal init, relying on setAppId (usage order does not matter)
+    * Sets the global shared ApplicationContext for OneSignal
+    * This is should be set from all OneSignal entry points
+    *   - BroadcastReceivers, Services, and Activities
+    * @param context - Context used by the Application of the app
+    */
+   public static void setAppContext(@NonNull Context context) {
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppContext(context) called!");
+
+      if (context == null) {
+         Log(LOG_LEVEL.WARN, "context is null, ignoring!");
+         return;
       }
+
+      boolean wasAppContextNull = (appContext == null);
+      appContext = context.getApplicationContext();
+      setupActivityLifecycleListener(wasAppContextNull);
+      setupPrivacyConsent(appContext);
+
+      OneSignal.onesignalLog(LOG_LEVEL.VERBOSE, "setAppContext(context) finished, checking if appId has been set before proceeding...");
+      if (appId == null) {
+         OneSignal.onesignalLog(LOG_LEVEL.WARN, "appContext set, but please call setAppId(appId) with a valid appId to complete OneSignal init!");
```

Comment:
Attempt to read `appId` from shared perfs when context is set. Wrapper SDK will call setContext first and setAppId won't be called until the wrapper env starts up

Reply:
Okay sounds good
Ill just do `appId = getSavedAppId();`
before the null check